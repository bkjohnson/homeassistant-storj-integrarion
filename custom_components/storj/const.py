"""Constants for the storj integration."""

DOMAIN = "storj"
CONF_ACCESS_GRANT = "access_grant"
CONF_BUCKET_NAME = "bucket_name"
