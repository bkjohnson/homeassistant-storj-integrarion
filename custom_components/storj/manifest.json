{
  "domain": "storj",
  "name": "Storj",
  "after_dependencies": ["backup"],
  "codeowners": ["@bkjohnson"],
  "config_flow": true,
  "dependencies": [],
  "documentation": "https://github.com/bkjohnson/homeassistant-storj-integration",
  "homekit": {},
  "iot_class": "cloud_polling",
  "issue_tracker": "https://github.com/bkjohnson/homeassistant-storj-integration/issues",
  "requirements": [],
  "ssdp": [],
  "version": "0.1.0",
  "zeroconf": []
}
