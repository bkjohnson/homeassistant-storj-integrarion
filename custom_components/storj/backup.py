"""Backup platform for the Storj integration."""

from __future__ import annotations

from collections.abc import AsyncIterator, Callable
import logging
from typing import Any
from pathlib import Path

from homeassistant.components.backup import AgentBackup, BackupAgent, BackupAgentError
from homeassistant.core import HomeAssistant, callback
from homeassistant.exceptions import HomeAssistantError

from . import DATA_BACKUP_AGENT_LISTENERS, StorjConfigEntry
from .const import DOMAIN

_LOGGER = logging.getLogger(__name__)


async def async_get_backup_agents(
    hass: HomeAssistant,
    **kwargs: Any,
) -> list[BackupAgent]:
    """Return a list of backup agents."""
    entries = hass.config_entries.async_loaded_entries(DOMAIN)
    return [StorjBackupAgent(hass, entry) for entry in entries]


@callback
def async_register_backup_agents_listener(
    hass: HomeAssistant,
    *,
    listener: Callable[[], None],
    **kwargs: Any,
) -> Callable[[], None]:
    """Register a listener to be called when agents are added or removed.
    :return: A function to unregister the listener.
    """
    hass.data.setdefault(DATA_BACKUP_AGENT_LISTENERS, []).append(listener)

    @callback
    def remove_listener() -> None:
        """Remove the listener."""
        hass.data[DATA_BACKUP_AGENT_LISTENERS].remove(listener)
        if not hass.data[DATA_BACKUP_AGENT_LISTENERS]:
            del hass.data[DATA_BACKUP_AGENT_LISTENERS]

    return remove_listener


class StorjBackupAgent(BackupAgent):
    """Storj backup agent."""

    domain = DOMAIN

    def __init__(self, hass: HomeAssistant, config_entry: StorjConfigEntry) -> None:
        """Initialize the cloud backup sync agent."""
        super().__init__()
        assert config_entry.unique_id
        self.hass = hass
        self.name = config_entry.title
        self.unique_id = config_entry.unique_id
        self._backup_dir = Path(hass.config.path("backups"))
        self._client = config_entry.runtime_data

    async def async_upload_backup(
        self,
        *,
        backup: AgentBackup,
        **kwargs: Any,
    ) -> None:
        """Upload a backup.
        :param backup: Metadata about the backup that should be uploaded.
        """
        try:
            await self._client.async_upload_backup(self._backup_dir, backup)
        except (HomeAssistantError, TimeoutError) as err:
            raise BackupAgentError(f"Failed to upload backup: {err}") from err

    async def async_list_backups(self, **kwargs: Any) -> list[AgentBackup]:
        """List backups."""
        try:
            return await self._client.async_list_backups()
        except (HomeAssistantError, TimeoutError) as err:
            raise BackupAgentError(f"Failed to list backups: {err}") from err

    async def async_get_backup(
        self,
        backup_id: str,
        **kwargs: Any,
    ) -> AgentBackup | None:
        """Return a backup."""
        backups = await self.async_list_backups()
        for backup in backups:
            if backup.backup_id == backup_id:
                return backup
        return None

    async def async_download_backup(
        self,
        backup_id: str,
        **kwargs: Any,
    ) -> AsyncIterator[bytes]:
        """Download a backup file.
        :param backup_id: The ID of the backup that was returned in async_list_backups.
        :return: An async iterator that yields bytes.
        """
        _LOGGER.debug("Downloading backup_id: %s", backup_id)
        # try:
        #     file_id = await self._client.async_get_backup_file_id(backup_id)

    async def async_delete_backup(
        self,
        backup_id: str,
        **kwargs: Any,
    ) -> None:
        """Delete a backup file.
        :param backup_id: The ID of the backup that was returned in async_list_backups.
        """
        _LOGGER.debug("Deleting backup_id: %s", backup_id)
        # file_id = await self._client.async_get_backup_file_id(backup_id)
