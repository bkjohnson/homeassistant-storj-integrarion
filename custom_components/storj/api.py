"""API for Home Assistant to interact with Storj."""

from __future__ import annotations

import asyncio
import logging

from homeassistant.components.backup import AgentBackup, suggested_filename

_LOGGER = logging.getLogger(__name__)


class StorjClient:
    """Client for Storj uplink CLI tool."""

    def __init__(
        self,
        ha_instance_id: str,
        bucket_name: str,
    ) -> None:
        """Initialize."""
        self._ha_instance_id = ha_instance_id
        self.bucket_name = bucket_name
        # self.satellite = satellite

    async def authenticate(self, access_grant: str) -> bool:
        """Test if we can authenticate with the host."""
        result = await asyncio.create_subprocess_exec(
            "uplink", "access", "import", "ha2", access_grant
        )
        await result.communicate()
        return result.returncode == 0

    async def async_upload_backup(
        self,
        backup_dir: str,
        backup: AgentBackup,
    ) -> None:
        """Upload a backup."""
        # backup_metadata = {
        #     "name": suggested_filename(backup),
        #     "description": json.dumps(backup.as_dict()),
        #     "properties": {
        #         "home_assistant": "backup",
        #         "instance_id": self._ha_instance_id,
        #         "backup_id": backup.backup_id,
        #     },
        # }
        _LOGGER.debug(
            "Uploading backup: %s as %s",
            backup.backup_id,
            suggested_filename(backup),
            # backup_metadata,
        )

        # TODO: Add metadata,
        backup_location = f"{backup_dir}/{suggested_filename(backup)}"
        result = await asyncio.create_subprocess_exec(
            "uplink", "cp", backup_location, f"sj://{self.bucket_name}"
        )
        await result.communicate()
        _LOGGER.debug("Uploaded backup: %s to '%s'", backup.backup_id, self.bucket_name)

    async def async_list_backups(self) -> list[AgentBackup]:
        """List the backups currently in the bucket."""
        # query = " and ".join(
        #     [
        #         "properties has { key='home_assistant' and value='backup' }",
        #         f"properties has {{ key='instance_id' and value='{self._ha_instance_id}' }}",
        #         "trashed=false",
        #     ]
        # )
        # res = await self._api.list_files(
        #     params={"q": query, "fields": "files(description)"}
        # )
        backups: list[AgentBackup] = []
        # for file in res["files"]:
        #     backup = AgentBackup.from_dict(json.loads(file["description"]))
        #     backups.append(backup)
        return backups

    # async def async_list_backups(self) -> None:
    #     """List the backups currently in the bucket."""
    #     _LOGGER.debug("TODO")

    async def async_delete_backup(self) -> None:
        """Delete a specified backup from the bucket."""
        _LOGGER.debug("TODO")

    async def async_download_backup(self) -> None:
        """Download a backup to the local system."""
        _LOGGER.debug("TODO")
