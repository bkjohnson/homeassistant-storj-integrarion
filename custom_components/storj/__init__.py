"""Custom integration to integrate Storj with Home Assistant as a backup location."""

from __future__ import annotations

from collections.abc import Callable

from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers import instance_id
from homeassistant.util.hass_dict import HassKey

from .api import StorjClient
from .const import DOMAIN, CONF_BUCKET_NAME

type StorjConfigEntry = ConfigEntry[StorjClient]

DATA_BACKUP_AGENT_LISTENERS: HassKey[list[Callable[[], None]]] = HassKey(
    f"{DOMAIN}.backup_agent_listeners"
)


async def async_setup_entry(hass: HomeAssistant, entry: StorjConfigEntry) -> bool:
    """Set up storj from a config entry."""

    # TODO 1. Create API instance
    # TODO 2. Validate the API connection (and authentication)
    # TODO 3. Store an API object for your platforms to access

    entry.runtime_data = StorjClient(
        await instance_id.async_get(hass), entry.data[CONF_BUCKET_NAME]
    )

    return True


async def async_unload_entry(hass: HomeAssistant, entry: StorjConfigEntry) -> bool:
    """Unload a config entry."""
    hass.loop.call_soon(_notify_backup_listeners, hass)
    return True


def _notify_backup_listeners(hass: HomeAssistant) -> None:
    for listener in hass.data.get(DATA_BACKUP_AGENT_LISTENERS, []):
        listener()
