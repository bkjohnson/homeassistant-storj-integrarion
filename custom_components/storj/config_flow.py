"""Config flow for the storj integration."""

from __future__ import annotations

import logging
from typing import Any

import voluptuous as vol

from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
from homeassistant.core import HomeAssistant
from homeassistant.exceptions import HomeAssistantError
from homeassistant.helpers import instance_id

from .api import StorjClient
from .const import CONF_ACCESS_GRANT, CONF_BUCKET_NAME, DOMAIN

_LOGGER = logging.getLogger(__name__)

# TODO adjust the data schema to the data that you need
STEP_USER_DATA_SCHEMA = vol.Schema(
    {
        vol.Required(CONF_ACCESS_GRANT): str,
        vol.Required(CONF_BUCKET_NAME): str,
    }
)


async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> dict[str, Any]:
    """Validate the user input allows us to connect.
    Data has the keys from STEP_USER_DATA_SCHEMA with values provided by the user.
    """
    client = StorjClient(await instance_id.async_get(hass), data[CONF_BUCKET_NAME])

    if not await client.authenticate(data[CONF_ACCESS_GRANT]):
        raise InvalidAuth

    # If you cannot connect:
    # throw CannotConnect
    # If the authentication is wrong:
    # InvalidAuth

    # Return info that you want to store in the config entry.
    return {"title": "Storj"}


class StorjConfigFlow(ConfigFlow, domain=DOMAIN):
    """Handle a config flow for storj."""

    VERSION = 1

    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the initial step."""
        errors: dict[str, str] = {}
        if user_input is not None:
            try:
                info = await validate_input(self.hass, user_input)
            except CannotConnect:
                errors["base"] = "cannot_connect"
            except InvalidAuth:
                errors["base"] = "invalid_auth"
            except Exception:
                _LOGGER.exception("Unexpected exception")
                errors["base"] = "unknown"
            else:
                await self.async_set_unique_id(user_input[CONF_ACCESS_GRANT])
                return self.async_create_entry(title=info["title"], data=user_input)

        return self.async_show_form(
            step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
        )


class CannotConnect(HomeAssistantError):
    """Error to indicate we cannot connect."""


class InvalidAuth(HomeAssistantError):
    """Error to indicate there is invalid auth."""
