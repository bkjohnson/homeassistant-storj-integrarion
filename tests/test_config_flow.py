"""Test Storj config flow."""

from unittest.mock import AsyncMock, patch

from homeassistant import config_entries
from custom_components.storj.config_flow import CannotConnect, InvalidAuth
from custom_components.storj.const import DOMAIN, CONF_ACCESS_GRANT, CONF_BUCKET_NAME
from syrupy.assertion import SnapshotAssertion
from homeassistant.core import HomeAssistant
from homeassistant.data_entry_flow import FlowResultType

from .conftest import mock_asyncio_subprocess_run


async def test_form(
    hass: HomeAssistant, mock_setup_entry: AsyncMock, snapshot: SnapshotAssertion
) -> None:
    """Test we get the form."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )
    assert result["type"] is FlowResultType.FORM
    assert result["errors"] == {}

    with mock_asyncio_subprocess_run(responses=iter([b""])) as subprocess_exec:
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_ACCESS_GRANT: "abc123xyz",
                CONF_BUCKET_NAME: "my-backups",
            },
        )
        await hass.async_block_till_done()
        assert snapshot() == subprocess_exec.mock_calls[0].args

    assert result["type"] is FlowResultType.CREATE_ENTRY
    assert result["title"] == "Storj"
    assert result["data"] == {
        CONF_ACCESS_GRANT: "abc123xyz",
        CONF_BUCKET_NAME: "my-backups",
    }
    assert len(mock_setup_entry.mock_calls) == 1


async def test_form_invalid_auth(
    hass: HomeAssistant, mock_setup_entry: AsyncMock
) -> None:
    """Test we handle invalid auth."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "custom_components.storj.api.StorjClient.authenticate",
        side_effect=InvalidAuth,
    ):
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_ACCESS_GRANT: "abc123xyz",
                CONF_BUCKET_NAME: "my-backups",
            },
        )

    assert result["type"] is FlowResultType.FORM
    assert result["errors"] == {"base": "invalid_auth"}

    # Make sure the config flow tests finish with either an
    # FlowResultType.CREATE_ENTRY or FlowResultType.ABORT so
    # we can show the config flow is able to recover from an error.
    with patch(
        "custom_components.storj.api.StorjClient.authenticate",
        return_value=True,
    ):
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_ACCESS_GRANT: "abc123xyz",
                CONF_BUCKET_NAME: "my-backups",
            },
        )
        await hass.async_block_till_done()

    assert result["type"] is FlowResultType.CREATE_ENTRY
    assert result["title"] == "Storj"
    assert result["data"] == {
        CONF_ACCESS_GRANT: "abc123xyz",
        CONF_BUCKET_NAME: "my-backups",
    }
    assert len(mock_setup_entry.mock_calls) == 1


async def test_form_cannot_connect(
    hass: HomeAssistant, mock_setup_entry: AsyncMock
) -> None:
    """Test we handle cannot connect error."""
    result = await hass.config_entries.flow.async_init(
        DOMAIN, context={"source": config_entries.SOURCE_USER}
    )

    with patch(
        "custom_components.storj.api.StorjClient.authenticate",
        side_effect=CannotConnect,
    ):
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_ACCESS_GRANT: "abc123xyz",
                CONF_BUCKET_NAME: "my-backups",
            },
        )

    assert result["type"] is FlowResultType.FORM
    assert result["errors"] == {"base": "cannot_connect"}

    # Make sure the config flow tests finish with either an
    # FlowResultType.CREATE_ENTRY or FlowResultType.ABORT so
    # we can show the config flow is able to recover from an error.

    with patch(
        "custom_components.storj.api.StorjClient.authenticate",
        return_value=True,
    ):
        result = await hass.config_entries.flow.async_configure(
            result["flow_id"],
            {
                CONF_ACCESS_GRANT: "abc123xyz",
                CONF_BUCKET_NAME: "my-backups",
            },
        )
        await hass.async_block_till_done()

    assert result["type"] is FlowResultType.CREATE_ENTRY
    assert result["title"] == "Storj"
    assert result["data"] == {
        CONF_ACCESS_GRANT: "abc123xyz",
        CONF_BUCKET_NAME: "my-backups",
    }
    assert len(mock_setup_entry.mock_calls) == 1
