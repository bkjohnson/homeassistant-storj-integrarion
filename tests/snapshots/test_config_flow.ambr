# serializer version: 1
# name: test_form
  tuple(
    'uplink',
    'access',
    'import',
    'ha2',
    'abc123xyz',
  )
# ---
